src/MemObject.cc

```diff
@@ -353,6 +353,12 @@ MemObject::policyLowestOffsetToKeep(bool swap) const
      */
     int64_t lowest_offset = lowestMemReaderOffset();
 
+    // XXX: Remove the last (Config.onoff.memory_cache_first-based) condition
+    // and update keepForLocalMemoryCache() accordingly. The caller wants to
+    // remove all local memory that is safe to remove. Honoring caching
+    // preferences is its responsibility. Our responsibility is safety. The
+    // situation was different when ff4b33f added that condition -- there was no
+    // keepInLocalMemory/keepForLocalMemoryCache() call guard back then.
     if (endOffset() < lowest_offset ||
             endOffset() - inmem_lo > (int64_t)Config.Store.maxInMemObjSize ||
             (swap && !Config.onoff.memory_cache_first))
```

Comment:
Do you agree with this XXX? This question has the lowest priority.

Reply:
It looks far from obvious that those keepInLocalMemory/keepForLocalMemoryCache() calls prior to our policyLowestOffsetToKeep() have already checked that Config.onoff.memory_cache_first is false. There are lots of checks in those methods, and the Config.onoff.memory_cache_first is one of them (and not the first one). Moreover, why are we concerned only with this check, and omit the second one (Config.Store.maxInMemObjSize-based)? the memory object size should have been checked in those methods too...

> and update keepForLocalMemoryCache() accordingly

Maybe after this intended updating the questions above will be resolved. However, until it is done, I cannot state that following this XXX suggestion would not change memory trimming behavior.

Reply:
> XXX: and update keepForLocalMemoryCache() accordingly

> Eduard: Maybe after this intended updating the questions above will be resolved

Yes, the "accordingly" bit is meant to imply that updated keepForLocalMemoryCache() should prevent unwanted[^1] and trigger desirable[^1] memory trimming. That method makes the decision to trim or to keep. Our method just tells the caller how much trimming is currently safe (from the current memory locking/use by Store point of view) if the decision to trim has been made.

[^1]: Unwanted or desirable from the Config.onoff.memory_cache_first point of view and from any other relevant caching policy points of view. There are indeed a lot of things to check when making that decision, and keepForLocalMemoryCache() may be missing some necessary checks right now.